src/lib.rs

```diff
@@ -110,7 +171,7 @@ macro_rules! from_impl {
     ($from:ty => $to:ty; $by:ident) => (
         impl<'a> From<$from> for UniCase<$to> {
             fn from(s: $from) -> Self {
-                UniCase(s.$by())
+                UniCase::new(s.$by())
```

Comment:
This could be quite a hidden cost in generic code if it will incur a full scan for a potentially large string (compared to how it should be cheap for failed equality tests). I think it would be better to require the user to opt-in to the potential ASCII optimisation.